lib/Backend/Lower.cpp

```diff
@@ -10382,6 +10375,168 @@ Lowerer::LowerEqualityBranch(IR::Instr* instr, IR::JnHelperMethod helper)
+            instr->InsertBefore(branch);
+        }
+        else
+        {
+            m_lowererMD.GenerateObjectTest(src1->AsRegOpnd(), instr, labelHelper);
+
+#if !FLOATVAR
+            IR::RegOpnd *src1TypeReg = IR::RegOpnd::New(TyMachReg, this->m_func);
+            Lowerer::InsertMove(src1TypeReg, IR::IndirOpnd::New(src1->AsRegOpnd(), Js::RecyclableObject::GetOffsetOfType(), TyMachReg, this->m_func), instr);
+
+            // MOV src1TypeIdReg, [src1TypeReg + offset(typeId)]
+            IR::RegOpnd *src1TypeIdReg = IR::RegOpnd::New(TyInt32, this->m_func);
+            Lowerer::InsertMove(src1TypeIdReg, IR::IndirOpnd::New(src1TypeReg, Js::Type::GetOffsetOfTypeId(), TyInt32, this->m_func), instr);
+
+            // CMP src1TypeIdReg, TypeIds_Number
+            // JEQ $helper
+            IR::IntConstOpnd *numberTypeId = IR::IntConstOpnd::New(Js::TypeIds_Number, TyInt32, this->m_func, true);
+            InsertCompareBranch(src1TypeIdReg, numberTypeId, Js::OpCode::BrEq_A, labelHelper, instr);
+#endif
+            //      CMP src1, src2                       - Ptr comparison
+            //      JEQ $branchSuccess
+            InsertCompareBranch(src1, src2, Js::OpCode::BrEq_A, labelBranchSuccess, instr);
+
+#if FLOATVAR
+            IR::RegOpnd *src1TypeReg = IR::RegOpnd::New(TyMachReg, this->m_func);
+            Lowerer::InsertMove(src1TypeReg, IR::IndirOpnd::New(src1->AsRegOpnd(), Js::RecyclableObject::GetOffsetOfType(), TyMachReg, this->m_func), instr);
+
+            // MOV src1TypeIdReg, [src1TypeReg + offset(typeId)]
+            IR::RegOpnd *src1TypeIdReg = IR::RegOpnd::New(TyInt32, this->m_func);
+            Lowerer::InsertMove(src1TypeIdReg, IR::IndirOpnd::New(src1TypeReg, Js::Type::GetOffsetOfTypeId(), TyInt32, this->m_func), instr);
+#endif
+            // CMP src1TypeIdReg, TypeIds_HostDispatch
+            // JLE $helper (le condition covers string, int64, uint64, hostdispatch, as well as undefined, null, boolean)
+            IR::IntConstOpnd *hostDispatchTypeId = IR::IntConstOpnd::New(Js::TypeIds_HostDispatch, TyInt32, this->m_func, true);
+            InsertCompareBranch(src1TypeIdReg, hostDispatchTypeId, Js::OpCode::BrLe_A, labelHelper, instr);
+
+            // CMP src1TypeIdReg, TypeIds_GlobalObject
+            // JE $helper
+            IR::IntConstOpnd *globalObjectTypeId = IR::IntConstOpnd::New(Js::TypeIds_GlobalObject, TyInt32, this->m_func, true);
+            InsertCompareBranch(src1TypeIdReg, globalObjectTypeId, Js::OpCode::BrEq_A, labelHelper, instr);
+
+            // TEST src1TypeReg->flags, TypeFlagMask_EngineExternal
+            // JE $helper
+
+            IR::Opnd *flags = IR::IndirOpnd::New(src1TypeReg, Js::Type::GetOffsetOfFlags(), TyInt8, this->m_func);
+            InsertTestBranch(flags, IR::IntConstOpnd::New(TypeFlagMask_EngineExternal, TyInt8, this->m_func), Js::OpCode::BrNeq_A, labelHelper, instr);
+
```

Comment:
@rajatd With this change, the helper path will have a fast path for objects that are not GlobalObject, HostDispatch or External in that order to simply have the pointer comparison for strict equality, instead of calling the C++ helper as before.